volume/guest_image.go

```diff
@@ -0,0 +1,217 @@
+	image, err := p.client.Pull(ctx, p.image, remoteOpts...)
+	if err != nil {
+		return err
+	}
+
+	p.cImage = image
+	return nil
+}
+
+func mountHostDirs(image string) oci.SpecOpts {
+	return func(ctx context.Context, client oci.Client, container *containers.Container, s *oci.Spec) error {
+		s.Mounts = append(s.Mounts,
+			specs.Mount{
+				Source:      vmBinDir,
+				Destination: containerBinDir,
+				Type:        "bind",
+				Options:     []string{"bind", "ro"},
+			},
+			specs.Mount{
+				Source:      vmVolumesDir,
+				Destination: containerVolumesDir,
+				Type:        "bind",
+				Options:     []string{"bind"},
+			},
+		)
+		return nil
+	}
+}
+
+// Copy files from the image by launching the container.
+func (p *GuestVolumeImageProvider) Copy(ctx context.Context, name string) error {
+	container, err := p.client.NewContainer(ctx,
+		name,
+		containerd.WithSnapshotter(p.config.snapshotter),
+		containerd.WithNewSnapshot(p.snapshot, p.cImage),
+		containerd.WithNewSpec(
+			firecrackeroci.WithVMID(p.vmID),
+			oci.WithProcessArgs(filepath.Join(containerBinDir, "volume-init")),
+			oci.WithProcessCwd("/"),
+			mountHostDirs(name),
+		),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create container %q with %q: %w", name, p.config.snapshotter, err)
+	}
+	defer container.Delete(ctx, containerd.WithSnapshotCleanup)
+
+	input := GuestVolumeImageInput{From: "/", To: filepath.Join(containerVolumesDir, name), Volumes: p.volumes}
+	b, err := json.Marshal(&input)
+	if err != nil {
+		return err
+	}
+
+	stdin := bytes.NewBuffer(b)
+	var stdout, stderr bytes.Buffer
+	task, err := container.NewTask(ctx, cio.NewCreator(cio.WithStreams(stdin, &stdout, os.Stderr)))
+	if err != nil {
+		return err
+	}
+	defer task.Delete(ctx)
+
+	exitCh, err := task.Wait(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = task.Start(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = task.CloseIO(ctx, containerd.WithStdinCloser)
+	if err != nil {
+		return err
+	}
+
+	select {
+	case status := <-exitCh:
+		if err := status.Error(); err != nil {
+			return err
+		}
+
+		_, err := task.Delete(ctx)
+		if err != nil {
+			return err
+		}
+
+		code := status.ExitCode()
+		if code != 0 {
+			return fmt.Errorf(
+				"failed to copy files from %q: stdout=%q, stderr=%q, code=%d",
+				p.image,
+				stdout.String(), stderr.String(), status.ExitCode(),
+			)
+		}
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	p.vmDir = filepath.Join(vmVolumesDir, name)
+	return nil
+}
+
+// CreateVolumesUnder creates volumes under the given directory.
+func (p *GuestVolumeImageProvider) CreateVolumesUnder(ctx context.Context, tempDir string) ([]*Volume, error) {
+	if p.vmDir == "" {
+		return nil, errors.New("call Copy() beforehand")
+	}
+
+	result := make([]*Volume, 0, len(p.volumes))
+	for _, path := range p.volumes {
+		result = append(result, &Volume{
+			vmPath:        filepath.Join(p.vmDir, path),
+			containerPath: path,
+		})
+	}
+	return result, nil
+}
```

Comment:
I probably should make this method specific to imageVolumeProvider. GuestVolumeImageProvider doesn't use tempDir.